feat: DAO SDK (#1925)
gnolang/gno: docs/how-to-guides/creating-dao.md

## Practical Implementation

### Setting Up Your Workspace

Sooo, let's create a new realm

```
git clone https://github.com/TERITORI/gno.git gno-dao-tutorial
cd gno-dao-tutorial
git checkout teritori-unified
mkdir examples/gno.land/r/demo/my_dao
```

### Creating the Voting Module

We will start by instantiating a voting module

1. **Initialize the Factory**

Modules instantiation uses the factory pattern in case the module needs to access the core

`examples/gno.land/r/demo/my_dao/my_dao.gno`
```go
package my_dao

import (
    "gno.land/p/demo/teritori/dao_interfaces"
)

func init() {
    votingModuleFactory := func(core dao_interfaces.IDAOCore) {
   
    }
}
```

2. **Instantiate the module**

`examples/gno.land/r/demo/my_dao/my_dao.gno`
```go
package my_dao

import (
    "gno.land/p/demo/teritori/dao_interfaces"
    "gno.land/p/demo/teritori/dao_voting_group" // <- new
)

func init() {
    var group *dao_voting_group.VotingGroup // <- new

    votingModuleFactory := func(core dao_interfaces.IDAOCore) {
        group = dao_voting_group.NewVotingGroup() // <- new
    }

> }

Reviewer: It's not crucial, but what do you think about marking changes in the code snippet like diff? 

For example, line that are added or modifed would be marked with a `+` sign, and lines that are deleted would be marked with a `-` sign. The downside in this case would be that we cannot use syntax highlighting.

Example:

```diff
import (
    "gno.land/p/demo/teritori/dao_interfaces"
+    "gno.land/p/demo/teritori/dao_voting_group"
)

func init() {
+    var group *dao_voting_group.VotingGroup

    votingModuleFactory := func(core dao_interfaces.IDAOCore) {
+        group = dao_voting_group.NewVotingGroup()
    }
}
```

Author: I prefer to keep syntax highlighting and pastability

---
id: creating-dao
---

# How to Create a DAO

## Overview

> This guide will show you how to write a simple [**DAO**](https://en.wikipedia.org/wiki/Decentralized_autonomous_organization) [realm](../concepts/realms.md) in [Gno](../concepts/gno-language.md). For actually deploying the Realm, please see the

Reviewer 2: ```suggestion
This guide will show you how to write a simple [**DAO**](https://en.wikipedia.org/wiki/Decentralized_autonomous_organization) [realm](../concepts/realms.md) in [Gno](../concepts/gno-language.md). For actually deploying the realm, please see the
```

> This guide will show you how to write a simple [**DAO**](https://en.wikipedia.org/wiki/Decentralized_autonomous_organization) [realm](../concepts/realms.md) in [Gno](../concepts/gno-language.md). For actually deploying the Realm, please see the

Reviewer 2: ```suggestion
This guide will show you how to write a simple [**DAO**](https://en.wikipedia.org/wiki/Decentralized_autonomous_organization) [realm](../concepts/realms.md) in [Gno](../concepts/gno-language.md). For actually deploying the realm, please see the
```

	VotingPowerAtHeight(address std.Address, height int64) (power uint64)
	TotalPowerAtHeight(height int64) uint64
}
```

There is only one implementation currently, `gno.land/p/demo/dao_maker/dao_voting_group`, providing a membership-based voting power definition.

### Proposal Modules

A proposal module (`gno.land/p/demo/dao_maker/dao_interfaces.IProposalModule`) is responsible for:
- Receiving proposals, the proposal type is defined by the module
- Managing the proposals lifecycle
- Tallying votes, the vote type is defined by the module and the associated voting power is queried from the voting module
- Executing proposals once they are passed

**Interface Definition:**
```go
type IProposalModule interface {
	Core() IDAOCore
	Info() ModuleInfo
	ConfigJSON() string
	Render(path string) string
	Execute(proposalID int)
	VoteJSON(proposalID int, voteJSON string)
	ProposeJSON(proposalJSON string) int
	ProposalsJSON(limit int, startAfter string, reverse bool) string
	ProposalJSON(proposalID int) string
}
```

There is only one implementation currently, `gno.land/p/demo/dao_maker/dao_proposal_single`, providing a yes/no/abstain vote model with quorum and threshold.

### Message handlers

> Proposals actions are encoded as objects implementing `gno.land/p/demo/dao_maker/dao_interfaces.ExecutableMessage`.

Reviewer 2: ```suggestion
Proposal actions are encoded as objects implementing `ExecutableMessage` found under `gno.land/p/demo/dao_maker/dao_interfaces`.
```

```go
type ExecutableMessage interface {
	ToJSON() *json.Node
	FromJSON(ast *json.Node)

	String() string
	Type() string
}
```

> They are deserialized and executed by message handlers implementing `gno.land/p/demo/dao_maker/dao_interfaces.MessageHandler`.

Reviewer 2: ```suggestion
They are unmarshalled and executed by message handlers implementing `gno.land/p/demo/dao_maker/dao_interfaces.MessageHandler`.
```
Marshal/unmarshal is a more "go way" to say serialize/deserialize.

```go
type MessageHandler interface {
	Execute(message ExecutableMessage)
	Instantiate() ExecutableMessage
	Type() string
}
```

Message handlers are registered at core creation and new message handlers can be registered via proposals to extend the DAO capabilities.

## Practical Implementation

> 

Reviewer 2: ```suggestion
In this section, we will showcase how to implement your own DAO using the `dao_maker` package suite.

```

### Setting Up Your Workspace

#### Setup the Tooling

To setup your tooling, see [Getting Started: Local Setup](../getting-started/local-setup.md).

#### Create a new Gno module

- Create a new directory and move into it: `mkdir my-gno-dao && cd my-gno-dao`

> - Initialize the gno module: `gno mod init gno.land/r/demo/my_dao`

Reviewer 2: ```suggestion
- Initialize the gno module: `gno mod init gno.land/r/<your_namespace>/my_dao`
```
